From 8631afc9b6caf0003667416b49c73c09475bd785 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Tue, 25 Jul 2023 15:30:36 +0800
Subject: [PATCH 14/53] [Huawei] mm/sharepool: Save nodemask in
 sp_alloc_context

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

1. Parse the node_id in sp_alloc_prepare(), because the node_id is
   used only for sp_alloc(it makes no scene for k2u).
2. Save nodemask in sp_alloc_context during sp_alloc.
3. Check the input device_id.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 65 +++++++++++++++++++++++++++----------------------
 1 file changed, 36 insertions(+), 29 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 2c1951e79a7a..49adac5d7321 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -1613,22 +1613,20 @@ static void insert_sp_area(struct sp_mapping *spm, struct sp_area *spa)
  */
 static struct sp_area *sp_alloc_area(unsigned long size, unsigned long flags,
 				     struct sp_group *spg, enum spa_type type,
-				     pid_t applier)
+				     pid_t applier, int node_id)
 {
+	int device_id;
 	struct sp_area *spa, *first, *err;
 	struct rb_node *n;
 	unsigned long vstart;
 	unsigned long vend;
 	unsigned long addr;
 	unsigned long size_align = ALIGN(size, PMD_SIZE); /* va aligned to 2M */
-	int device_id, node_id;
 	struct sp_mapping *mapping;
 
 	device_id = sp_flags_device_id(flags);
-	node_id = flags & SP_SPEC_NODE_ID ? sp_flags_node_id(flags) : device_id;
-
-	if (!is_online_node_id(node_id)) {
-		pr_err_ratelimited("invalid numa node id %d\n", node_id);
+	if (device_id < 0 || device_id >= MAX_DEVID) {
+		pr_err("invalid device id %d\n", device_id);
 		return ERR_PTR(-EINVAL);
 	}
 
@@ -2112,6 +2110,8 @@ struct sp_alloc_context {
 	unsigned long size;
 	unsigned long size_aligned;
 	unsigned long sp_flags;
+	nodemask_t *nodemask;
+	int preferred_node_id;
 	bool have_mbind;
 	enum spa_type type;
 };
@@ -2120,9 +2120,21 @@ static int sp_alloc_prepare(unsigned long size, unsigned long sp_flags,
 	int spg_id, struct sp_alloc_context *ac)
 {
 	struct sp_group *spg;
+	int device_id, node_id;
+
+	if (!sp_is_enabled())
+		return -EOPNOTSUPP;
 
 	check_interrupt_context();
 
+	device_id = sp_flags_device_id(sp_flags);
+	node_id = sp_flags & SP_SPEC_NODE_ID ? sp_flags_node_id(sp_flags) : device_id;
+	if (!is_online_node_id(node_id)) {
+		pr_err_ratelimited("invalid numa node id %d\n", node_id);
+		return -EINVAL;
+	}
+	ac->preferred_node_id = node_id;
+
 	if (current->flags & PF_KTHREAD) {
 		pr_err_ratelimited("allocation failed, task is kthread\n");
 		return -EINVAL;
@@ -2211,8 +2223,7 @@ static long sp_mbind(struct mm_struct *mm, unsigned long start, unsigned long le
 
 #define SP_SKIP_ERR 1
 static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
-	unsigned long prot, struct sp_alloc_context *ac,
-	nodemask_t *nodemask)
+				    unsigned long prot, struct sp_alloc_context *ac)
 {
 	int ret;
 	unsigned long mmap_addr;
@@ -2236,10 +2247,10 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 	up_write(&mm->mmap_lock);
 
 	if (!ac->have_mbind) {
-		ret = sp_mbind(mm, spa->va_start, spa->real_size, nodemask);
+		ret = sp_mbind(mm, spa->va_start, spa->real_size, ac->nodemask);
 		if (ret < 0) {
 			pr_err("cannot bind the memory range to node[%*pbl], err:%d\n",
-					nodemask_pr_args(nodemask), ret);
+					nodemask_pr_args(ac->nodemask), ret);
 			down_write(&mm->mmap_lock);
 			do_munmap(mm, mmap_addr, spa_size(spa), NULL);
 			up_write(&mm->mmap_lock);
@@ -2269,26 +2280,17 @@ static int __sp_alloc_mmap_populate(struct mm_struct *mm, struct sp_area *spa,
 	return ret;
 }
 
-static int sp_alloc_mmap_populate(struct sp_area *spa,
-				  struct sp_alloc_context *ac,
-				  nodemask_t *nodemask)
+static int sp_alloc_mmap_populate(struct sp_area *spa, struct sp_alloc_context *ac)
 {
 	int ret = -EINVAL;
 	int mmap_ret = 0;
 	struct mm_struct *mm;
 	struct sp_group_node *spg_node;
-	nodemask_t __nodemask;
-
-	if (!nodemask) { /* mg_sp_alloc */
-		nodes_clear(__nodemask);
-		node_set(spa->preferred_node_id, __nodemask);
-	} else /* mg_sp_alloc_nodemask */
-		__nodemask = *nodemask;
 
 	/* create mapping for each process in the group */
 	list_for_each_entry(spg_node, &spa->spg->proc_head, proc_node) {
 		mm = spg_node->master->mm;
-		mmap_ret = __sp_alloc_mmap_populate(mm, spa, spg_node->prot, ac, &__nodemask);
+		mmap_ret = __sp_alloc_mmap_populate(mm, spa, spg_node->prot, ac);
 		if (mmap_ret) {
 			/*
 			 * Goto fallback procedure upon ERR_VALUE,
@@ -2342,22 +2344,27 @@ static void sp_alloc_finish(int result, struct sp_area *spa,
 }
 
 void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg_id,
-		nodemask_t *nodemask)
+			     nodemask_t *nodemask)
 {
-	struct sp_area *spa = NULL;
 	int ret = 0;
+	nodemask_t __nodemask;
+	struct sp_area *spa = NULL;
 	struct sp_alloc_context ac;
 
-	if (!sp_is_enabled())
-		return ERR_PTR(-EOPNOTSUPP);
-
 	ret = sp_alloc_prepare(size, sp_flags, spg_id, &ac);
 	if (ret)
 		return ERR_PTR(ret);
 
+	if (!nodemask) { /* mg_sp_alloc */
+		nodes_clear(__nodemask);
+		node_set(ac.preferred_node_id, __nodemask);
+		ac.nodemask = &__nodemask;
+	} else /* mg_sp_alloc_nodemask */
+		ac.nodemask = nodemask;
+
 try_again:
 	spa = sp_alloc_area(ac.size_aligned, ac.sp_flags, ac.spg,
-			    ac.type, current->tgid);
+			    ac.type, current->tgid, ac.preferred_node_id);
 	if (IS_ERR(spa)) {
 		pr_err_ratelimited("alloc spa failed in allocation(potential no enough virtual memory when -75): %ld\n",
 			PTR_ERR(spa));
@@ -2365,7 +2372,7 @@ void *__mg_sp_alloc_nodemask(unsigned long size, unsigned long sp_flags, int spg
 		goto out;
 	}
 
-	ret = sp_alloc_mmap_populate(spa, &ac, nodemask);
+	ret = sp_alloc_mmap_populate(spa, &ac);
 	if (ret == -ENOMEM && sp_alloc_fallback(spa, &ac))
 		goto try_again;
 
@@ -2521,7 +2528,7 @@ static void *sp_make_share_kva_to_spg(struct sp_k2u_context *kc, struct sp_group
 	unsigned long ret_addr = -ENODEV;
 
 	down_read(&spg->rw_lock);
-	spa = sp_alloc_area(kc->size_aligned, kc->sp_flags, spg, kc->type, current->tgid);
+	spa = sp_alloc_area(kc->size_aligned, kc->sp_flags, spg, kc->type, current->tgid, 0);
 	if (IS_ERR(spa)) {
 		up_read(&spg->rw_lock);
 		pr_err("alloc spa failed in k2u_spg (potential no enough virtual memory when -75): %ld\n",
-- 
2.17.1

